File: algorithm-java/src/main/java/com/algorithm/funny/algorithm/CoinCollect.java
package com.algorithm.funny.algorithm;

/**
 * @author clx
 */
public class CoinCollect {

	/**
	 * @param coins
	 * @param amount
	 * @return
	 */
	public static int coinCollect(int[] coins, int amount) {
		if (amount < 0) {
			return -1;
		}
		if (amount == 0) {
			return 0;
		}
		int[] dp = new int[amount + 1];
		for (int idx = 1; idx < dp.length; idx++) {
			dp[idx] = idx;
		}
		for (int idx = 0; idx < dp.length; idx++) {
			for (int coin : coins) {
				if (idx - coin < 0) {
					continue;
				}
				dp[idx] = Math.min(dp[idx], 1 + dp[idx - coin]);
			}
		}
		return (dp[amount] == amount + 1) ? -1 : dp[amount];
	}
}

File: algorithm-python/algorithm/coin_collect.py
# -*- coding: utf-8 -*-

class CoinCollect:

    @staticmethod
    def coin_change(coins: list[int], amount: int):
        def dp(n):
            if amount < 0:
                return -1
            if amount == 0:
                return 0
            res = float('INF')
            for coin in coins:
                sub = dp(n - coin)
                if sub == -1:
                    continue
                res = min(res, 1 + sub)
            return res if res != float('INF') else -1

        return dp(amount)

    @staticmethod
    def coin_change_map(coins: list[int], amount: int):
        table = dict()

        def dp(n):
            if n in table:
                return table[n]
            if n < 0:
                return -1
            if n == 0:
                return 0
            res = float('INF')
            for coin in coins:
                sub = dp(n - coin)
                if sub == -1:
                    continue
                res = min(res, 1 + sub)
            table[n] = res if res != float('INF') else -1
            return table[n]

        return dp(amount)

    @staticmethod
    def coin_collect(coins: list[int], amount: int) -> int:
        return
